keps/sig-api-machinery/NNNN-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,739 @@
+
+#### Story 2
+
+As a Kubernetes developer I'd like to be able to develop HA applications and controllers without requiring external components.
+
+#### Story 3
+
+As a Kubernetes architect I'd like the cluster components to be completely autonomous.
+
+### Notes/Constraints/Caveats (Optional)
+
+Network load balancing is complicated, there are many edge cases and there is no universal solution, the requirements also change and evolve with the time (see HTTP1.1 vs HTTP2).
+
+Client-go is also the library used by all components, any change or regression there can have a great impact on the project.
+
+The implementation of this feature should not modify any of current behaviors and completely pluggable, so users can opt-in safely and the project can be sure there will not create any regression.
+
+### Risks and Mitigations
+
+The feature will be opt-in, users can disable it and keep having the same behavior.
+
+To avoid possible security issues, this feature will require HTTP2 and HTTPS in order to work.
+
+## Design Details
+
+There are two problems we have to solve:
+
+- get a list of available control-plane nodes
+- client capability to be able to connect to multiple control-planes
+
+### List of available control-plane nodes
+
+This problems can be solved in two different ways:
+
+1. Configuring on the client the list of available control-plane nodes
+
+This can be done extending the client-go configuration with an additional field that allows users to include a list of alternative control-plane nodes that the client can use.
+
+`k8s.io/client-go/rest/config.go`
+```go
+type Config struct {
+	// Host must be a host string, a host:port pair, or a URL to the base of the apiserver.
+	// If a URL is given then the (optional) Path of that URL represents a prefix that must
+	// be appended to all request URIs used to access the apiserver. This allows a frontend
+	// proxy to easily relocate all of the apiserver endpoints.
+	Host string
+  // AlternativeHosts must be a comma separated list if hosts, host:port pairs or URLs to the base of
+  // different apiservers. The client can use any of them to access the apiserver.
+	AlternativeHosts string
+```
+
+2. RFC7838 HTTP Alternative Services
+
+This specification defines a new concept in HTTP, "Alternative Services", that allows an origin server to nominate additional means of interacting with it on the network.
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, per example:
```

Comment:
These aren't guaranteed to be accessible to clients that will be getting them back.  It could fail based on "you cannot access this" and it could fail based on "you're contacting a different 10.0.0.2 because you're on a different network".  The kube-apiserver cannot not know the network connectivity of the client.

Reply:
I'm operating under the assumption that this is opt-in and users can only use it for in-cluster communication, where the networking model ensure that endpoints are reachable by any other pod and host in the cluster.
If there are some edge cases it can be always filtered,  apiservers can check the original ip of the request per example

Reply:
```suggestion
The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, this requires that the users of this feature belong to the cluster network. TODO define a way of filtering the headers in the answer, per example, by source IP, certificate name, ... 

```